Introduce OptionSet to updateLayout

rwlbuis opened the pull request:
#### 9ed0cc30d357eb9673342ff8bd1d8143485034c8
<pre>
Introduce OptionSet to updateLayout
<a href="https://bugs.webkit.org/show_bug.cgi?id=261105">https://bugs.webkit.org/show_bug.cgi?id=261105</a>

Reviewed by Tim Nguyen.

Introduce OptionSet to updateLayout to be able to specify its
behaviour, i.e. whether to run post lasyout tasks and/or ignore stylesheets.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateLayoutIgnorePendingStylesheets):
(WebCore::Document::updateLayout):
* Source/WebCore/dom/Document.h:
(WebCore::Document::updateLayout):
(WebCore::Document::updateLayoutIgnorePendingStylesheets):
* Source/WebCore/html/HTMLPlugInElement.cpp:
(WebCore::HTMLPlugInElement::renderWidgetLoadingPlugin const):
* Source/WebCore/page/VisualViewport.cpp:
(WebCore::VisualViewport::updateFrameLayout const):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::absoluteLineRectFromPoint):
(WebCore::Internals::boundingBox):
(WebCore::Internals::layoutViewportRect):
(WebCore::Internals::visualViewportRect):
(WebCore::Internals::lineIndexAfterPageBreak):
(WebCore::Internals::recentSearches):
(WebCore::Internals::nodesFromRect const):
(WebCore::Internals::layerIDForElement):
(WebCore:: const):
(WebCore::Internals::scrollingStateTreeAsText const):
(WebCore::Internals::scrollingTreeAsText const):
(WebCore::Internals::haveScrollingTree const):
(WebCore::Internals::setElementUsesDisplayListDrawing):
(WebCore::Internals::setElementTracksDisplayListReplay):
(WebCore::Internals::displayListForElement):
(WebCore::Internals::replayDisplayListForElement):
(WebCore::Internals::cachedGlyphDisplayListsForTextNode):
(WebCore::Internals::updateLayoutIgnorePendingStylesheetsAndRunPostLayoutTasks):
(WebCore::Internals::isSelectPopupVisible):
(WebCore::Internals::pageOverlayLayerTreeAsText const):
* Source/WebCore/testing/Internals.mm:
(WebCore::Internals::rangeForDictionaryLookupAtLocation):
* Source/WebKit/WebProcess/InjectedBundle/API/mac/WKDOMNode.mm:
(-[WKDOMNode textRects]):
* Source/WebKit/WebProcess/InjectedBundle/API/mac/WKDOMRange.mm:
(-[WKDOMRange textRects]):
* Source/WebKit/WebProcess/WebPage/ViewGestureGeometryCollector.cpp:
(WebKit::ViewGestureGeometryCollector::computeTextLegibilityScales):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::elementBoundsInFrame):
(WebKit::WebPage::requestDocumentEditingContext):
(WebKit::WebPage::focusTextInputContextAndPlaceCaret):
* Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.mm:
(WebKit::inlineVideoFrame):
* Source/WebKitLegacy/mac/DOM/DOM.mm:
(-[DOMNode boundingBox]):
(-[DOMNode absoluteQuadAndInsideFixedPosition:]):
(-[DOMNode boundingBoxUsingTransforms]):
(-[DOMNode lineBoxQuads]):
(-[DOMNode innerFrameQuad]):
(-[DOMNode textRects]):
(-[DOMRange boundingBox]):
(-[DOMRange textRects]):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView updateLayoutIgnorePendingStyleSheets]):

Canonical link: <a href="https://commits.webkit.org/268241@main">https://commits.webkit.org/268241@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2375f4782bacd7e4440b7fec907ed986e77acdab

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17357 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17681 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18190 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19143 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16240 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17553 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20957 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17823 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/18413 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17561 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17899 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15093 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19961 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15139 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15790 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22457 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16138 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15959 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20284 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16537 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14034 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15683 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4557 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20053 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16373 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/dom/Document.h:
```diff
@@ -333,6 +333,10 @@ enum class DimensionsCheck : uint8_t {
 };
 
+enum class LayoutOptions : uint8_t {
+    RunPostLayoutTasksSynchronously = 1 << 0,
```
Should the "IgnorePendingStylesheets" behavior also be part of the options?

rwlbuis replied:
I like that idea! It makes the OptionSet a lot more useful in my opinion.
In my update I still have updateLayoutIgnorePendingStylesheets in order to not have to update the many callsites, however I have no problem updating them here or in a followup. But also let's see if it goes green....

rwlbuis committed:
Introduce OptionSet to updateLayout
https://bugs.webkit.org/show_bug.cgi?id=261105

Reviewed by Tim Nguyen.

Introduce OptionSet to updateLayout to be able to specify its
behaviour, i.e. whether to run post lasyout tasks and/or ignore stylesheets.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateLayoutIgnorePendingStylesheets):
(WebCore::Document::updateLayout):
* Source/WebCore/dom/Document.h:
(WebCore::Document::updateLayout):
(WebCore::Document::updateLayoutIgnorePendingStylesheets):
* Source/WebCore/html/HTMLPlugInElement.cpp:
(WebCore::HTMLPlugInElement::renderWidgetLoadingPlugin const):
* Source/WebCore/page/VisualViewport.cpp:
(WebCore::VisualViewport::updateFrameLayout const):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::absoluteLineRectFromPoint):
(WebCore::Internals::boundingBox):
(WebCore::Internals::layoutViewportRect):
(WebCore::Internals::visualViewportRect):
(WebCore::Internals::lineIndexAfterPageBreak):
(WebCore::Internals::recentSearches):
(WebCore::Internals::nodesFromRect const):
(WebCore::Internals::layerIDForElement):
(WebCore:: const):
(WebCore::Internals::scrollingStateTreeAsText const):
(WebCore::Internals::scrollingTreeAsText const):
(WebCore::Internals::haveScrollingTree const):
(WebCore::Internals::setElementUsesDisplayListDrawing):
(WebCore::Internals::setElementTracksDisplayListReplay):
(WebCore::Internals::displayListForElement):
(WebCore::Internals::replayDisplayListForElement):
(WebCore::Internals::cachedGlyphDisplayListsForTextNode):
(WebCore::Internals::updateLayoutIgnorePendingStylesheetsAndRunPostLayoutTasks):
(WebCore::Internals::isSelectPopupVisible):
(WebCore::Internals::pageOverlayLayerTreeAsText const):
* Source/WebCore/testing/Internals.mm:
(WebCore::Internals::rangeForDictionaryLookupAtLocation):
* Source/WebKit/WebProcess/InjectedBundle/API/mac/WKDOMNode.mm:
(-[WKDOMNode textRects]):
* Source/WebKit/WebProcess/InjectedBundle/API/mac/WKDOMRange.mm:
(-[WKDOMRange textRects]):
* Source/WebKit/WebProcess/WebPage/ViewGestureGeometryCollector.cpp:
(WebKit::ViewGestureGeometryCollector::computeTextLegibilityScales):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::elementBoundsInFrame):
(WebKit::WebPage::requestDocumentEditingContext):
(WebKit::WebPage::focusTextInputContextAndPlaceCaret):
* Source/WebKit/WebProcess/cocoa/VideoFullscreenManager.mm:
(WebKit::inlineVideoFrame):
* Source/WebKitLegacy/mac/DOM/DOM.mm:
(-[DOMNode boundingBox]):
(-[DOMNode absoluteQuadAndInsideFixedPosition:]):
(-[DOMNode boundingBoxUsingTransforms]):
(-[DOMNode lineBoxQuads]):
(-[DOMNode innerFrameQuad]):
(-[DOMNode textRects]):
(-[DOMRange boundingBox]):
(-[DOMRange textRects]):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView updateLayoutIgnorePendingStyleSheets]):

Canonical link: https://commits.webkit.org/268241@main
